pkg/ottl/functions.go

```diff
@@ -32,65 +55,67 @@ func (p *Parser[K]) newFunctionCall(inv invocation) (Expr[K], error) {
 	if !ok {
 		return Expr[K]{}, fmt.Errorf("undefined function %v", inv.Function)
 	}
-	args, err := p.buildArgs(inv, reflect.TypeOf(f))
+	args := f.CreateEmptyArguments()
+
+	// Pointer values are necessary to fulfill the Go reflection
+	// settability requirements. Non-pointer values are not
+	// modifiable through reflection.
+	if reflect.TypeOf(args).Kind() != reflect.Pointer {
+		return Expr[K]{}, fmt.Errorf("factory for %s must return a pointer to an Arguments value in its CreateEmptyArguments method", inv.Function)
+	}
+
+	err := p.buildArgs(inv, reflect.ValueOf(args).Elem())
 	if err != nil {
 		return Expr[K]{}, fmt.Errorf("error while parsing arguments for call to '%v': %w", inv.Function, err)
 	}
 
-	returnVals := reflect.ValueOf(f).Call(args)
-
-	if returnVals[1].IsNil() {
-		err = nil
-	} else {
-		err = returnVals[1].Interface().(error)
+	fn, err := f.CreateFunction(context.Background(), p.telemetrySettings, args)
+	if err != nil {
+		return Expr[K]{}, fmt.Errorf("couldn't create function")
 	}
 
-	return Expr[K]{exprFunc: returnVals[0].Interface().(ExprFunc[K])}, err
+	return Expr[K]{exprFunc: fn}, err
 }
 
-func (p *Parser[K]) buildArgs(inv invocation, fType reflect.Type) ([]reflect.Value, error) {
-	var args []reflect.Value
-	// Some function arguments may be intended to take values from the calling processor
-	// instead of being passed by the caller of the OTTL function, so we have to keep
-	// track of the index of the argument passed within the DSL.
-	// e.g. TelemetrySettings, which is provided by the processor to the OTTL Parser struct.
-	DSLArgumentIndex := 0
-	for i := 0; i < fType.NumIn(); i++ {
-		argType := fType.In(i)
+func (p *Parser[K]) buildArgs(inv invocation, argsVal reflect.Value) error {
+	if len(inv.Arguments) > argsVal.NumField() {
```

Comment:
Can we add a similar check for `len(inv.Arguments) < argsVal.NumField()`?

Reply:
Good call, I've set it to be an equality check.